server/src/main/java/org/elasticsearch/cluster/coordination/MutableMasterHistory.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point. This object is updated
+ * if and when the cluster state changes with a new master node.
+ */
+public class MutableMasterHistory implements MasterHistory, ClusterStateListener {
+    volatile List<TimeAndMaster> masterHistory;
+    private final Object mutex;
+    Supplier<Long> nowSupplier; // Can be changed for testing
+
+    public MutableMasterHistory(ThreadPool threadPool, ClusterService clusterService) {
+        this.masterHistory = new ArrayList<>();
+        this.mutex = new Object();
+        this.nowSupplier = threadPool::relativeTimeInMillis;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null || currentMaster.equals(previousMaster) == false || masterHistory.isEmpty()) {
+            masterHistory.add(new TimeAndMaster(nowSupplier.get(), currentMaster));
+        }
+        removeOldMasterHistory();
```

Comment:
Should we only purge if we're modifying the history? (give we're also purging on every read)

Reply:
Good catch. Changing that now.